cub/test/c2h/generators.cuh

```diff
@@ -110,5 +145,34 @@ template <typename T>
 thrust::device_vector<T>
 gen_uniform_offsets(seed_t seed, T total_elements, T min_segment_size, T max_segment_size);
 
-} // c2h
+/**
+ * @brief Generates key-segment ranges from an offsets-array like the one given by
+ * `gen_uniform_offset`.
+ */
+template <typename key_t, typename offset_t>
+thrust::device_vector<key_t>
+get_key_segments(const thrust::device_vector<offset_t> &segment_offsets, offset_t total_elements)
+{
+  // Constant iterator returning '1' for any index
+  key_t non_zero_constant{};
+  detail::init_non_zero_constant(non_zero_constant);
+  auto const_one_it = thrust::make_constant_iterator(non_zero_constant);
+
+  // Iterator of segment offsets (skipping the first segment offset of '0')
+  auto segment_offset_it = thrust::next(segment_offsets.cbegin());
+
+  // Prepare keys array, scattering '1' to segment offsets
+  thrust::device_vector<key_t> key_segments(total_elements);
+  auto num_segments = segment_offsets.size() - 1;
+  thrust::scatter_if(const_one_it,
+                     const_one_it + num_segments,
+                     segment_offset_it,
+                     segment_offset_it,
+                     key_segments.begin(),
+                     detail::le_comparator_op<offset_t>{total_elements - 1});
+
+  thrust::inclusive_scan(key_segments.cbegin(), key_segments.cend(), key_segments.begin());
```

Comment:
Is this approach better than run-length-decode in `gen_power_law_key_segments` (nvbench helper)? From what I gather, it enforces presence of `operator+` in custom type. If we test an algorithm with user-provided operator, I'd prefer our test type to provide the minimal set of methods. Otherwise, we risk relying on the extra methods in our code. If you preferred this approach to the run-length-decode one because of performance benefits, could you share what's the speedup in question? If it significantly affects the test runtime, we could expose it with a different name and rely on it for tests that require `operator+` anyways.

Reply:
After offline discussions, we came to the conclusion that a run-length decoding-based approach imposes fewer requirements on the key type.